rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java

```diff
@@ -75,8 +98,190 @@ public NodesLockRestorationManager apply(RMCore rmCore) {
+        return recoverNodes;
+    }
+
+    private void recoverDeployedNodeSourceIfNeeded(NodeSourceData nodeSourceDescriptor, String nodeSourceName) {
+        try {
+            logger.info("Recover node source " + nodeSourceName);
+            // retrieve node source status
+            boolean deployNodeSource = false;
+            if (nodeSourceDescriptor.getStatus().equals(NodeSourceStatus.NODES_DEPLOYED)) {
+                deployNodeSource = true;
+            }
+            // reset node source status
+            nodeSourceDescriptor.setStatus(NodeSourceStatus.NODES_UNDEPLOYED);
+            rmCore.createNotActiveNodeSource(nodeSourceDescriptor);
+            // repeat node source status
+            if (deployNodeSource) {
+                rmCore.deployNodeSource(nodeSourceName);
+            }
+        } catch (Throwable t) {
+            logger.error("Failed to recover node source " + nodeSourceName, t);
+            rmCore.removeNodeSource(nodeSourceName);
+            rmCore.getDbManager().removeNodeSource(nodeSourceName);
+        }
+    }
+
+    private void logPersistedNodeSourceInfo(Collection<NodeSourceData> nodeSources) {
+        if (nodeSources.isEmpty()) {
+            logger.info("No node source found in database");
+        } else {
+            if (nodeSources.size() < 10) {
+                logger.info("Node sources found in database: " + Arrays.toString(nodeSources.toArray()));
+            } else {
+                logger.info("Number of node sources found in database: " + nodeSources.size());
+            }
+        }
+    }
+
+    public void recoverNodes(NodeSource nodeSource) {
+        logger.info(START_TO_RECOVER_NODES); // this log line is important for performance tests
+        int lookUpTimeout = PAResourceManagerProperties.RM_NODELOOKUP_TIMEOUT.getValueAsInt();
+        String nodeSourceName = nodeSource.getName();
+
+        makeSureNodeSourceHasNoNode(nodeSource, nodeSourceName);
+
+        Collection<RMNodeData> nodesData = rmCore.getDbManager().getNodesByNodeSource(nodeSourceName);
+        logger.info("Number of nodes found in database for node source " + nodeSourceName + ": " + nodesData.size());
+
+        Map<NodeState, Integer> nodeStates = new HashMap<>();
+        int totalEligibleRecoveredNodes = 0;
+
+        // for each node found in database, try to lookup node or recover it
+        // as down node
+        for (RMNodeData rmNodeData : nodesData) {
+            String nodeUrl = rmNodeData.getNodeUrl();
+            RMNode rmnode = null;
+
+            Node node = tryToLookupNode(nodeSource, lookUpTimeout, nodeUrl);
+
+            if (node != null) {
+                rmnode = recoverNodeInternally(nodeSource, rmNodeData, nodeUrl, node);
+                nodesLockRestorationManager.handle(rmnode, rmNodeData.getProvider());
+                Integer nbNodesInState = nodeStates.get(rmnode.getState());
+                int newNbNodesInState = nbNodesInState == null ? 1 : nbNodesInState + 1;
```

Comment:
nbNodesInState and newNbNodesInState are very similar. I find it hard to not confuse them, maybe there is a different naming possible? 

nbNodesInState == null ? 1 : nbNodesInState + 1;

Personally I think it is better to write the if as a full if () {}else{} .

Something like this:
```
int newNodesInState = 1;
if( nbNodesInState 1= null) { 
newNodesInState = nbNodesInState + 1
}
```

Reply:
you could initialize the nbNodesInState map with 0 values. This way, the code to increment would be much nicer.

Reply:
done